pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go

```diff
@@ -0,0 +1,228 @@
+
+func (r *REST) List(ctx context.Context, options *metainternal.ListOptions) (runtime.Object, error) {
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sanitizedListOpts, err := userFieldSelectorToListOptions(options, ctxUserName)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenListGenericObj, err := r.accessTokenStorage.List(ctx, sanitizedListOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenList, ok := tokenListGenericObj.(*oauthapi.OAuthAccessTokenList)
+	if !ok {
+		return nil, errors.NewInternalError(fmt.Errorf("failed to convert generic accesstoken LIST result to its typed version"))
+	}
+
+	return oauthAccessTokenListToUserOAuthAccessTokenList(tokenList, ctxUserName), nil
+}
+
+func (r *REST) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenGenericObj, err := r.accessTokenStorage.Get(ctx, name, options)
+	if err != nil {
+		return nil, err
+	}
+
+	token := tokenGenericObj.(*oauthapi.OAuthAccessToken)
+	if token == nil {
+		return nil, errors.NewInternalError(fmt.Errorf("failed to convert generic accesstoken GET result to its typed version"))
+	}
+
+	if isValidUserToken(token, ctxUserName) {
+		return nil, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+	}
+
+	return (*oauthapi.UserOAuthAccessToken)(token), nil
+}
+
+func (r *REST) Watch(ctx context.Context, options *metainternal.ListOptions) (watch.Interface, error) {
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sanitizedListOpts, err := userFieldSelectorToListOptions(options, ctxUserName)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenListWatcher, err := r.accessTokenStorage.Watch(ctx, sanitizedListOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	watcher := newOAuthAccessTokenWatcher(tokenListWatcher, ctxUserName)
+	go watcher.Watch(ctx)
+	return watcher, nil
+}
+
+func (r *REST) ConvertToTable(ctx context.Context, object runtime.Object, tableOptions runtime.Object) (*metav1.Table, error) {
+	return r.tableConvertor.ConvertToTable(ctx, object, tableOptions)
+}
+
+func (r *REST) NamespaceScoped() bool {
+	return false
+}
+
+func (r *REST) Delete(ctx context.Context, name string, validateFunc rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return &metav1.Status{Status: metav1.StatusFailure}, false, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+	}
+
+	var deletedRV string
+	if options.Preconditions != nil && options.Preconditions.ResourceVersion != nil {
+		deletedRV = *options.Preconditions.ResourceVersion
+	}
+
+	deletedObj, err := r.Get(ctx, name, &metav1.GetOptions{TypeMeta: options.TypeMeta, ResourceVersion: deletedRV})
+	if err != nil {
+		return &metav1.Status{Status: metav1.StatusFailure}, false, err
```

Comment:
why not nil as object?

Reply:
I was not sure how exactly this should be handled and the docs of the `GracefulDeleter` made me think this was the right way, but reading the code of the generic `Store`, it seems that `nil` would be the right option.